node/src/tasks/user_sends_file.rs

```diff
@@ -201,211 +204,233 @@ where
                                     peer_id,
-                                    e
+                                    file_metadata.fingerprint
                                 );
-                                break;
                             }
+                            break;
                         }
-                    }
-
-                    // Clear the batch for next iteration
-                    current_batch.clear();
-                    current_batch_size = 0;
-                }
+                        Err(RequestError::RequestFailure(RequestFailure::Refused))
+                            if retry_attempts < 3 =>
+                        {
+                            warn!(
+                                target: LOG_TARGET,
+                                "Final batch upload rejected by peer {:?}, retrying... (attempt {})",
+                                peer_id,
+                                retry_attempts + 1
+                            );
+                            retry_attempts += 1;
+
+                            // Wait for a short time before retrying
+                            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                        }
+                        Err(RequestError::RequestFailure(RequestFailure::Refused)) => {
+                            // If MSP doesn't receive file, the burden of downloading the file will be on the MSP.
 
-                // Add chunk to current batch
-                current_batch.push(ChunkId::new(chunk_id));
-                current_batch_size += chunk_size as usize;
-
-                // If this is the last chunk, send the final batch
-                if chunk_id == chunk_count - 1 && !current_batch.is_empty() {
-                    debug!(
-                        target: LOG_TARGET,
-                        "Sending final batch of {} chunks (total size: {} bytes) for file {:?} to peer {:?}",
-                        current_batch.len(),
-                        current_batch_size,
-                        file_key,
-                        peer_id
-                    );
-
-                    // Generate proof for the final batch
-                    let proof = match self
-                        .storage_hub_handler
-                        .file_storage
-                        .read()
-                        .await
-                        .generate_proof(&file_key, &current_batch)
-                    {
-                        Ok(proof) => proof,
+                            return Err(anyhow::anyhow!("Failed to send file {:?}", file_key));
+                        }
                         Err(e) => {
                             return Err(anyhow::anyhow!(
-                                "Failed to generate proof for final batch of file {:?}\n Error: {:?}",
-                                file_key,
+                                "Failed to upload final batch to peer {:?} (Error: {:?})",
```

Comment:
```suggestion
                                "Unexpected error while trying to upload final batch to peer {:?} \n    Error: {:?}",
```

I think this error message is clearer, given that it an error happened for which we didn't have a way to handle it. Also, the line break to print the `Error` helps readability.